In [ ]:
import numpy as np
import pandas as pd
import random as rd

# New Section

In [ ]:
dir(np)

['ALLOW_THREADS',
 'BUFSIZE',
 'CLIP',
 'DataSource',
 'ERR_CALL',
 'ERR_DEFAULT',
 'ERR_IGNORE',
 'ERR_LOG',
 'ERR_PRINT',
 'ERR_RAISE',
 'ERR_WARN',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'False_',
 'Inf',
 'Infinity',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'NAN',
 'NINF',
 'NZERO',
 'NaN',
 'PINF',
 'PZERO',
 'RAISE',
 'RankWarning',
 'SHIFT_DIVIDEBYZERO',
 'SHIFT_INVALID',
 'SHIFT_OVERFLOW',
 'SHIFT_UNDERFLOW',
 'ScalarType',
 'True_',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'WRAP',
 '_CopyMode',
 '_NoValue',
 '_UFUNC_API',
 '__NUMPY_SETUP__',
 '__all__',
 '__builtins__',
 '__cached__',
 '__config__',
 '__deprecated_attrs__',
 '__dir__',
 '__doc__',
 '__expired_functions__',
 '__file__',
 '__former_attrs__',
 '__future_scalars__',
 '__getattr__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_add_newdoc_ufunc',
 '_builtins',
 '_core',
 '_distributor_init',


In [ ]:
a = [[1,2,3,5],[3,4,5,7]]
print(type(a))
b = np.array(a)
print(type(b))
print(b.ndim)
print(b.shape)
print(b.size)

<class 'list'>
<class 'numpy.ndarray'>
2
(2, 4)
8


In [ ]:
c=np.array([1,2,3],ndmin=5)
print(c)
print(c.ndim)

[[[[[1 2 3]]]]]
5


In [ ]:
d=np.arange(10)
print(d)
e=d.reshape(5,2) # This only works if the dimentions match
e

[0 1 2 3 4 5 6 7 8 9]


array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7],
       [8, 9]])

In [ ]:
arr1 = np.array([1,2,3, True, False])
arr1

array([1, 2, 3, 1, 0])

In [ ]:
arr2 = np.array(['a','b','c','d',1,2,1.22])
arr2

array(['a', 'b', 'c', 'd', '1', '2', '1.22'], dtype='<U32')

In [ ]:
a=np.array([1,2,3])
print(a.dtype)
a.astype('float')
print(a.dtype)

int64
int64


In [ ]:
arr2[0:-2]

array(['a', 'b', 'c', 'd', '1'], dtype='<U32')

In [ ]:
testarr = np.array([1,2,3,4,'a','b'])
testarr=testarr.reshape(3,2) # Dont forget assignments lol
testarr.shape

(3, 2)

In [ ]:
a1 = np.array([[1,2,3],[4,5,6]])
a2 = np.array([[10,20,30],[40,50,60]])
# print(a1)
# print(a2)
np.concatenate((a1,a2),axis=1)
xxx = np.append(a1,a2,axis=0)
print(xxx)

[[ 1  2  3]
 [ 4  5  6]
 [10 20 30]
 [40 50 60]]


In [ ]:
np.delete(xxx[0],1)

array([ 1,  3, 10, 20, 30])

In [ ]:
#a=np.array([1,2,3,4,5,6,7,8,9,10])
a= np.arange(1,11)
print(a.min())
print(a.max())
print(a.std())
print(a.var())
print(a.mean())
np.median(a)

1
10
2.8722813232690143
8.25
5.5


5.5

In [ ]:
cw = np.arange(1,10)
cw=cw.reshape(3,3)
calcmin =[]
for i in range(3):
  x = cw[0][i]+cw[1][i]+cw[2][i]
  x=x/3
  calcmin.append(x)
print(cw)
print(calcmin)
print(cw.mean(axis=0))
cw.T

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[4.0, 5.0, 6.0]
[4. 5. 6.]


array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [ ]:
np.random.seed(123)
a = np.random.randint(1,11,size=10)
a

array([ 3,  3,  7,  2,  4, 10,  7,  2,  1,  2])

In [ ]:
v1 = np.array(np.random.randint(1,4,size=3))
v2 = np.array(np.random.randint(1,4,size=3))

In [ ]:
print(v1)
print(v2)
dotprod  = np.dot(v1,v2)
crossprod = np.cross(v1,v2)
print(dotprod)
print(crossprod)

[3 2 1]
[3 1 2]
13
[ 3 -3 -3]


In [ ]:
dicts = {'aaaa':10000,'bbbb':20000,'cccc':30000,'dddd':2134}
yy=pd.Series(dicts)

In [ ]:
ss = pd.Series(data=[1024,1234,1243,1521],index=['a','b','c','d'],name="Lol")
ss

a    1024
b    1234
c    1243
d    1521
Name: Lol, dtype: int64

In [ ]:
ss+yy

a      NaN
aaaa   NaN
b      NaN
bbbb   NaN
c      NaN
cccc   NaN
d      NaN
dddd   NaN
dtype: float64

In [ ]:
data={
    'NAME':['HARI','RAM','SITA','GITA'],
    'GENDER':['M','M','F','F'],
    'AGE':[11,22,33,44],
    'SALARY':[10024,22540,15684,15647]
}
df= pd.DataFrame(data)
print(df)

   NAME GENDER  AGE  SALARY
0  HARI      M   11   10024
1   RAM      M   22   22540
2  SITA      F   33   15684
3  GITA      F   44   15647


In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   NAME    4 non-null      object
 1   GENDER  4 non-null      object
 2   AGE     4 non-null      int64 
 3   SALARY  4 non-null      int64 
dtypes: int64(2), object(2)
memory usage: 256.0+ bytes
None


In [ ]:
df.describe()

AGE        SALARY
count   4.000000      4.000000
mean   27.500000  15973.750000
std    14.200939   5122.040113
min    11.000000  10024.000000
25%    19.250000  14241.250000
50%    27.500000  15665.500000
75%    35.750000  17398.000000
max    44.000000  22540.000000

In [ ]:
df.groupby('GENDER').describe()

AGE                                                 SALARY           \
       count  mean       std   min    25%   50%    75%   max  count     mean   
GENDER                                                                         
F        2.0  38.5  7.778175  33.0  35.75  38.5  41.25  44.0    2.0  15665.5   
M        2.0  16.5  7.778175  11.0  13.75  16.5  19.25  22.0    2.0  16282.0   

                                                                    
                std      min       25%      50%       75%      max  
GENDER                                                              
F         26.162951  15647.0  15656.25  15665.5  15674.75  15684.0  
M       8850.148473  10024.0  13153.00  16282.0  19411.00  22540.0

In [ ]:
df.groupby('GENDER').mean(numeric_only=True)

AGE   SALARY
GENDER               
F       38.5  15665.5
M       16.5  16282.0

# HOME WORK


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
hw = pd.read_csv("data_studentlist.csv")

In [ ]:
hw.columns

Index(['name', 'gender', 'age', 'grade', 'absence', 'bloodtype', 'height',
       'weight'],
      dtype='object')

In [ ]:
hw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17 entries, 0 to 16
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   name       17 non-null     object 
 1   gender     17 non-null     object 
 2   age        17 non-null     int64  
 3   grade      17 non-null     int64  
 4   absence    17 non-null     object 
 5   bloodtype  17 non-null     object 
 6   height     17 non-null     float64
 7   weight     17 non-null     float64
dtypes: float64(2), int64(2), object(4)
memory usage: 1.2+ KB


In [ ]:
dir(hw)
hw.groupby("gender").describe()

age                                                     grade  \
       count       mean       std   min   25%   50%    75%   max count   
gender                                                                   
F        7.0  21.857143  1.069045  20.0  21.5  22.0  22.50  23.0   7.0   
M       10.0  22.700000  1.159502  21.0  22.0  23.0  23.75  24.0  10.0   

                  ... height        weight                                    \
            mean  ...    75%    max  count       mean       std   min    25%   
gender            ...                                                          
F       1.714286  ...  173.1  176.9    7.0  50.442857  3.907624  45.2  47.40   
M       2.700000  ...  177.9  182.1   10.0  68.500000  9.505437  55.3  62.05   

                           
         50%    75%   max  
gender                     
F       52.0  53.05  55.0  
M       66.2  74.40  85.7  

[2 rows x 32 columns]

QN 1> F:M is 7:10

In [ ]:
hw.groupby("gender").mean(numeric_only=True)

age     grade      height     weight
gender                                            
F       21.857143  1.714286  166.642857  50.442857
M       22.700000  2.700000  172.410000  68.500000

QN 2> average height of Male:172.41 cm, and Female:166.64

In [ ]:
hw.groupby("bloodtype").describe()

age                                                     grade  \
          count       mean       std   min    25%   50%   75%   max count   
bloodtype                                                                   
A           4.0  21.250000  1.258306  20.0  20.75  21.0  21.5  23.0   4.0   
AB          3.0  22.333333  0.577350  22.0  22.00  22.0  22.5  23.0   3.0   
B           5.0  23.000000  1.000000  22.0  22.00  23.0  24.0  24.0   5.0   
O           5.0  22.600000  1.140175  21.0  22.00  23.0  23.0  24.0   5.0   

                     ...   height        weight                          \
               mean  ...      75%    max  count   mean        std   min   
bloodtype            ...                                                  
A          1.000000  ...  171.125  176.9    4.0  55.50   5.201923  49.5   
AB         2.333333  ...  181.050  182.1    3.0  71.50  16.768721  53.0   
B          3.000000  ...  176.100  178.5    5.0  64.20  12.827899  45.2   
O          2.600000  ...  167.100  176.1    5.0  56.12   8.995388  45.3   

                                        
              25%    50%     75%   max  
bloodtype                               
A          53.625  55.15  57.025  62.2  
AB         64.400  75.80  80.750  85.7  
B          61.300  64.20  70.200  80.1  
O          52.000  53.10  62.000  68.2  

[4 rows x 32 columns]

QN 3> Most common bloodtypes are B and O with a count of 5 each

In [ ]:
hw["height"]= hw["height"]/100
hw['height']

0     1.653
1     1.701
2     1.750
3     1.821
4     1.680
5     1.620
6     1.552
7     1.769
8     1.785
9     1.761
10    1.671
11    1.800
12    1.622
13    1.761
14    1.582
15    1.686
16    1.692
Name: height, dtype: float64

QN4 > Converted to meter

In [ ]:
hw = hw.sort_values(by=["gender","bloodtype"])
hw

name gender  age  grade absence bloodtype  height  weight
4    Clara Rodriquez      F   20      1       Y         A   1.680    49.5
7    Margareth Jones      F   23      1       N         A   1.769    55.0
1     Sarah O'Donnel      F   22      2       N        AB   1.701    53.0
14      Linda Carter      F   22      2       N         B   1.582    45.2
5   Jennifer Lorentz      F   21      2       N         O   1.620    52.0
6        Susan Clark      F   22      1       N         O   1.552    45.3
13  Rebecah Anderson      F   23      3       N         O   1.761    53.1
12      Eddy Johnson      M   21      1       N         A   1.622    55.3
16       Andrew King      M   21      1       N         A   1.692    62.2
3       David Hassel      M   23      3       N        AB   1.821    85.7
11      John Matsuda      M   22      2       N        AB   1.800    75.8
2       Brian Martin      M   24      4       N         B   1.750    80.1
8       John Bertsch      M   23      3       N         B   1.785    64.2
9     Jake Timmerman      M   22      2       N         B   1.761    61.3
15    Richard Swayze      M   24      4       Y         B   1.686    70.2
0      Jared Diamond      M   23      3       Y         O   1.653    68.2
10     Joshua Connor      M   24      4       Y         O   1.671    62.0

QN5 > Sort by gender and blood type

In [ ]:
pivot = pd.DataFrame()
pivot["bloodtype"] = hw["bloodtype"]
pivot

bloodtype
4          A
7          A
1         AB
14         B
5          O
6          O
13         O
12         A
16         A
3         AB
11        AB
2          B
8          B
9          B
15         B
0          O
10         O

In [ ]:
hw['BMI'] = hw['weight'] / (hw['height'] ** 2)

print(hw[['name', 'height', 'weight', 'BMI']])

                name  height  weight        BMI
4    Clara Rodriquez   1.680    49.5  17.538265
7    Margareth Jones   1.769    55.0  17.575473
1     Sarah O'Donnel   1.701    53.0  18.317544
14      Linda Carter   1.582    45.2  18.060321
5   Jennifer Lorentz   1.620    52.0  19.814053
6        Susan Clark   1.552    45.3  18.806794
13  Rebecah Anderson   1.761    53.1  17.122840
12      Eddy Johnson   1.622    55.3  21.019551
16       Andrew King   1.692    62.2  21.726495
3       David Hassel   1.821    85.7  25.844071
11      John Matsuda   1.800    75.8  23.395062
2       Brian Martin   1.750    80.1  26.155102
8       John Bertsch   1.785    64.2  20.149236
9     Jake Timmerman   1.761    61.3  19.767046
15    Richard Swayze   1.686    70.2  24.695736
0      Jared Diamond   1.653    68.2  24.959660
10     Joshua Connor   1.671    62.0  22.204387
